src/search/passes/transfer_to_knn.h

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+#pragma once
+
+#include <memory>
+
+#include "search/ir.h"
+#include "search/ir_pass.h"
+#include "search/ir_plan.h"
+
+namespace kqir {
+
+struct TransferSortByToKnnExpr : Visitor {
+  std::unique_ptr<Node> Visit(std::unique_ptr<SearchExpr> node) override {
+    node = Node::MustAs<SearchExpr>(Visitor::Visit(std::move(node)));
+
+    if (node->sort_by && node->sort_by->IsVectorField() && node->limit) {
+      std::vector<std::unique_ptr<QueryExpr>> exprs;
+      auto knn_expr = std::make_unique<VectorKnnExpr>(Node::MustAs<FieldRef>(node->sort_by->TakeFieldRef()),
+                                                      Node::MustAs<VectorLiteral>(node->sort_by->TakeVectorLiteral()),
+                                                      node->limit->Count());
+      if (auto b = Node::As<BoolLiteral>(std::move(node->query_expr))) {
+        if (b->val) exprs.push_back(std::move(knn_expr));
+      } else {
+        exprs.push_back(std::move(node->query_expr));
+        exprs.push_back(std::move(knn_expr));
+      }
```

Comment:
Seems here we cannot guarantee that the semantics of LIMIT clause is satisfied, e.g.
```sql
SELECT * FROM idx WHERE num > 1 ORDER BY vec <-> [...] LIMIT 10;
```
for this SQL statement, users want 10 records where `num > 1` (unless there's less than 10 records satisfying `num > 1`), 

but after transformation it's `(KNN(10) vec <-> [...]) AND num > 1`, which implies the number of records may be less than 10 even if there's more than 10 records satisfying `num > 1`, since it equals to filtering `num > 1` from `KNN(10)` query (firstly get 10 records from KNN search and then filter records of `num <= 1` out of these 10).

Reply:
And in the RediSearch query part, I think it's same, i.e. we cannot simply treat prefilter KNN search as AND expressions since it requries a specific number of records (`k`) -- and it breaks the semantics of `AND`.

Reply:
Thanks for pointing it out! Will prohibit hybrid query from this PR then. Also open to any suggestions from the community here <3

Reply:
Ready for review again!